server/irmaserver/sessions.go

```diff
@@ -72,45 +81,107 @@ type memorySessionStore struct {
 	client    map[irma.ClientToken]*session
 }
 
+type redisSessionStore struct {
+	client *redis.Client
+	locker *redislock.Client
+	conf   *server.Configuration
+}
+
+type RedisError struct {
+	err error
+}
+
+func (err *RedisError) Error() string {
+	return fmt.Sprintf("redis error: %s", err.err)
+}
+
+type UnknownSessionError struct {
+	requestorToken irma.RequestorToken
+	clientToken    irma.ClientToken
+}
+
+func (err *UnknownSessionError) Error() string {
+	if err.requestorToken != "" {
+		return fmt.Sprintf("session result requested of unknown session %s", err.requestorToken)
+	} else {
+		return fmt.Sprintf("session result requested of unknown session with clientToken %s", err.clientToken)
+	}
+}
+
+const (
+	maxLockLifetime            = 500 * time.Millisecond // After this the Redis lock self-deletes, preventing a deadlock
+	minLockRetryTime           = 30 * time.Millisecond
+	maxLockRetryTime           = 2 * time.Second
+	requestorTokenLookupPrefix = "token:"
+	clientTokenLookupPrefix    = "session:"
+	lockPrefix                 = "lock:"
+)
+
 var (
 	minProtocolVersion = irma.NewVersion(2, 4)
 	maxProtocolVersion = irma.NewVersion(2, 8)
 
 	minFrontendProtocolVersion = irma.NewVersion(1, 0)
 	maxFrontendProtocolVersion = irma.NewVersion(1, 1)
+
+	lockingRetryOptions = &redislock.Options{RetryStrategy: redislock.ExponentialBackoff(minLockRetryTime, maxLockRetryTime)}
 )
 
-func (s *memorySessionStore) get(t irma.RequestorToken) *session {
+func (s *memorySessionStore) get(t irma.RequestorToken) (*session, error) {
 	s.RLock()
 	defer s.RUnlock()
-	return s.requestor[t]
+	if s.requestor[t] != nil {
+		return s.requestor[t], nil
```

Comment:
Nitpich: this performs a lookup in the `s.requestor` map twice, which seems needlessly inefficient; wouldn't it be nicer to use a variable? Same thing in `clientGet()`.